%\VignetteIndexEntry{hiCLIP_data_analysis_1 Vignette}
%\VignetteKeyword{high-throughput DNA sequncing}
%\VignetteKeyword{RNA structures}
%\VignettePackage{hiCLIP_data_analysis_1}
\documentclass[12pt]{article}
\usepackage{hyperref}
\usepackage{url}

\newcommand{\textprime}{\textsuperscript{\ensuremath{\prime}}}

\textwidth=6.2in
\textheight=8.5in
\oddsidemargin=0.2in
\evensidemargin=0.2in
\headheight=0in
\headsep=0in

\begin{document}
\SweaveOpts{concordance=TRUE}

\title{Appendix 4/12}
%\author{Yoichiro Sugimoto and Jernej Ule}
\date{\today}
\maketitle

The appendix 4 performs hiCLIP data analysis. For this purpose, we introduce a S4 Class, Hybrid GRangesList class, which is the extension of GRangesList object from GenomicRanges package (from Bioconductor) as well as the associated Methods, which can be used for hiCLIP data analysis.

\tableofcontents

\section{Packages used}

<<Setting directory, echo=FALSE>>=
ptm <- proc.time()
options(width=60)
args <- commandArgs(trailingOnly = TRUE) 
hiCLIP_index <- grepl("hiCLIPdir=", args) 
hiCLIP.dir <- as.character(gsub("hiCLIPdir=", "", args[hiCLIP_index])) 
Rdir_index <- grepl("Rdir=", args) 
R.dir <- as.character(gsub("Rdir=", "", args[Rdir_index])) 
pythondir_index <- grepl("pythondir=", args) 
python.dir <- as.character(gsub("pythondir=", "", args[pythondir_index])) 

lib.path <- paste(R.dir, "/library/", sep = "")
.libPaths(lib.path)
@

Following packages are loaded.
<<set path to package, echo = TRUE>>=
library(plyr)
library(GenomicRanges)
library(ShortRead)
library(ggplot2)
@

\section{Import hybrid data}
<<source R codes, echo = FALSE>>=
sourceDir <- function(path, trace = TRUE, ...) {
    for (nm in list.files(path, pattern = "\\.[RrSsQq]$")) {
       source(file.path(path, nm), ...)
    }
}

S4dir <- paste(hiCLIP.dir, "/R/S4", sep = "")
sourceDir(S4dir)

functions_hotfactor.dir <- 
  paste(hiCLIP.dir, "/R/functions/plot_functions.R", sep = "")
source(functions_hotfactor.dir)
@

newHybridGRL takes bed-like file and sam-like file as input, both are generated by the previous steps of the hiCLIP analysis pipeline. The output is HybridGRL object, which is the extension of GRangesList object for hiCLIP data. First, in order to analyze the RNA source of hybrid reads in an unbiased manner, total hybrid reads (without considering random barcode) are analyzed. bed file from low RNase, high RNase and ligation minus condition are \texttt{total.low.bed.file}, \texttt{total.high.bed.file}, \texttt{nc.bed.file} respectively while sam file from those are \texttt{low.sam.file}, \texttt{high.sam.file} and \texttt{nc.sam.file}. 

<<specify path to row mapped data, echo = FALSE>>=
total.low.bed.file <- paste(hiCLIP.dir, "/results/mapping/hiCLIP/hybrid/bed/hybrid_DOX_LigPlus-barcode_for_hiClip-total.tab", sep = "")
total.high.bed.file <- paste(hiCLIP.dir, "/results/mapping/hiCLIP/hybrid/bed/hybrid_DOX_LigPlusHighRNase-barcode_for_hiClip-total.tab", sep = "")
total.nc.bed.file <- paste(hiCLIP.dir, "/results/mapping/hiCLIP/hybrid/bed/hybrid_DOX_LigMinus-barcode_for_hiClip-total.tab", sep = "")

low.sam.file <- paste(hiCLIP.dir, "/results/mapping/hiCLIP/hybrid/sam/hybrid_DOX_LigPlus.sam", sep = "")
high.sam.file <- paste(hiCLIP.dir, "/results/mapping/hiCLIP/hybrid/sam/hybrid_DOX_LigPlusHighRNase.sam", sep = "")
nc.sam.file <- paste(hiCLIP.dir, "/results/mapping/hiCLIP/hybrid/sam/hybrid_DOX_LigMinus.sam", sep = "")
@

<<Import Hybrid bed and sam file>>=
total.low.grL <- newHybridGRL(total.low.bed.file, low.sam.file)
total.high.grL <- newHybridGRL(total.high.bed.file, high.sam.file)
total.nc.grL <- newHybridGRL(total.nc.bed.file, nc.sam.file)
@

\section{Basic statistics of hybrid reads}
The number of valid hybrid reads (both arms' RNA source are uniquely annotated) is calculated by \texttt{readNumber()}.
<<Return read number for valid hybrid reads, echo=TRUE>>=
readNumber(total.low.grL)
readNumber(total.high.grL)
readNumber(total.nc.grL)
@

HybridGRL objects can be merged by \texttt{mergeHybrid()}.
<<Merge Hybrid reads, echo=TRUE>>=
total.merged.grl <- mergeHybrid(total.low.grL, total.high.grL)
@

\section{Analysis of the RNA source of hybrid reads}
Category and annotation of hybrid reads are added to HybridGRL object by \texttt{annotateHybrid()}.

<<Load files essential for annotation, echo=FALSE>>=
# t.gtf -> gr.tc.longest.mRNA
# mRNA.with.intron -> gr.mRNA.with.intron

gr.tc.longest.mRNA.Rdata <- paste(hiCLIP.dir, "/data/processed/Rdata/gr_tc_longest_mRNA.Rdata", sep = "")
load(gr.tc.longest.mRNA.Rdata)

gr.mRNA.with.intron.Rdata <- paste(hiCLIP.dir, "/data/processed/Rdata/gr_mRNA_with_intron.Rdata", sep = "")
load(gr.mRNA.with.intron.Rdata)

gr.intron <- gr.mRNA.with.intron[elementMetadata(gr.mRNA.with.intron)$annot == "intron"]

annotation.tab.file <- paste(hiCLIP.dir, "/data/processed/annotation/gene_annotations.tab", sep = "")
annotation.df <- read.table(annotation.tab.file, stringsAsFactors = FALSE, header = TRUE)
@

<<Annotate Hybrid reads, echo=TRUE>>=
total.merged.grl <- annotateHybrid(
  total.merged.grl, annotation.df, gr.tc.longest.mRNA, gr.mRNA.with.intron
  )
@

These data is plotted by \texttt{plotHybridRNASource()} (Figure~\ref{fig:Plot_hybrid_reads}).
<<Plot_hybrid_reads, echo=TRUE, fig=TRUE, include=FALSE>>=
plotHybridRNASource(total.merged.grl)
@

\begin{figure}[htbt]
\centering
\includegraphics[width=.6\linewidth]{hiCLIP_data_analysis_1-Plot_hybrid_reads}
\caption{RNA source of hybrid reads}
\label{fig:Plot_hybrid_reads}
\end{figure}


To test the validity of hybrid reads mapped to rRNAs, the free energy was compared with that of the random permutation control.

<<RNAhybrid_rRNA_preparation, echo=FALSE>>=
RNAhybrid.dir <- paste(hiCLIP.dir, "/results/manuscript/RNAhybrid", sep = "")
rRNA.fa.file <- paste(hiCLIP.dir, "/data/unprocessed/fasta/rRNA.fa", sep = "")
RNAhybrid.rRNA.dir <- paste(RNAhybrid.dir, "/rRNA", sep = "")
@

<<RNAhybridnalysis_for_rRNAs, echo=TRUE, fig=TRUE, include=FALSE>>=

rRNAs.hgrl <- select.rRNAsHybrid(total.merged.grl)

sr.rRNAs <- readFasta(rRNA.fa.file)

gr.rRNAs <- GRanges(seqnames = as.character(id(sr.rRNAs)),
                          ranges = IRanges(
                            start = 1,
                            width = width(sr.rRNAs)
                            ),
                          strand = strand("+"),
                          score = 1,
                          rname = 1:length(sr.rRNAs)
                      )

RNAhybridAnalysis(
  rRNAs.hgrl, 
  all.gr = gr.rRNAs, 
  fasta = sr.rRNAs, 
  out.dir = RNAhybrid.rRNA.dir, 
  hiCdir = hiCLIP.dir, 
  pdir = python.dir
  )
@

\begin{figure}[htbp]
\centering
\includegraphics[width=.6\linewidth]{hiCLIP_data_analysis_1-RNAhybridnalysis_for_rRNAs}
\caption{The folding energy of hybrid reads mapped to rRNAs}
\label{fig:RNAhybridnalysis_for_rRNAs}
\end{figure}

\section{Analysis of RNA structures with hybrid reads}

Analysis of RNA structure is performed with hybrid reads after PCR duplicate removal using the random barcode. 

<<specify path to cDNA counts bed file, echo = FALSE>>=
low.bed.file <- paste(hiCLIP.dir, "/results/mapping/hiCLIP/hybrid/bed/hybrid_DOX_LigPlus-barcode_for_hiClip-collapsed.tab", sep = "")
high.bed.file <- paste(hiCLIP.dir, "/results/mapping/hiCLIP/hybrid/bed/hybrid_DOX_LigPlusHighRNase-barcode_for_hiClip-collapsed.tab", sep = "")
@


<<Import Hybrid bed and sam file, echo=TRUE>>=
low.grL <- newHybridGRL(low.bed.file, low.sam.file)
high.grL <- newHybridGRL(high.bed.file, high.sam.file)
merged.grL <- mergeHybrid(low.grL, high.grL)
@

Hybrid reads uniquely mapped to genes after excluding reads from rRNAs, tRNAs, rDNAs, and mitochondrial genome are selected by \texttt{selectGeneMappedHybrid()}. 

<<Select Gene mapped hybrid reads, echo=TRUE>>=
merged.grL <- annotateHybrid(
  merged.grL, annotation.df, gr.tc.longest.mRNA, gr.mRNA.with.intron
  )
gene.grL <- selectGeneMappedHybrid(merged.grL)
@

The cDNA counts of hybrid reads uniquely mapped to genes are 

<<Return read number for valid hybrid reads, echo=TRUE>>=
readNumber(gene.grL)
@

The ratio of inter- or intra-molecular hybrid can be calculated by \texttt{plotInterVsIntra()} (Figure~\ref{fig:Plot_intra_vs_inter}).

<<Plot_intra_vs_inter, echo=TRUE, fig=TRUE, include=FALSE>>=
plotInterVsIntra(gene.grL)
@

\begin{figure}[htbt]
\centering
\includegraphics[width=.6\linewidth]{hiCLIP_data_analysis_1-Plot_intra_vs_inter}
\caption{Inter- vs Intra-molecular RNA structures}
\label{fig:Plot_intra_vs_inter}
\end{figure}

\section{Analysis of inter-molecular RNA strctures.}

Inter-molecular RNA hybrids are selected by \texttt{selectHybridType()} (Figure~\ref{fig:Select_inter_molecular_hybrids}).

<<Select_inter_molecular_hybrids, echo=TRUE, fig=TRUE, include=FALSE>>=
inter.gene.grl <- selectHybridType(gene.grL, "inter")
plotHybridRNASource(inter.gene.grl)
@

\begin{figure}[htbp]
\centering
\includegraphics[width=.6\linewidth]{hiCLIP_data_analysis_1-Select_inter_molecular_hybrids}
\caption{The RNA source of hybrid reads from inter-molecular interactions.}
\label{fig:Select_inter_molecular_hybrids}
\end{figure}

These statistics are compared with those from intra-molecular RNA hybrids (Figure~\ref{fig:Select_intra_molecular_hybrids}).

<<Select_intra_molecular_hybrids, echo=TRUE, fig=TRUE, include=FALSE>>=
intra.gene.grl <- selectHybridType(gene.grL, "intra")
plotHybridRNASource(intra.gene.grl)
@

<<Save_intra_molecular_hybrids, echo=FALSE>>=
intra.gene.grl.Rdata <- paste(hiCLIP.dir, "/data/processed/Rdata/intra.gene.grl.Rdata", sep = "")
save(intra.gene.grl, file = intra.gene.grl.Rdata) 
@


\begin{figure}[htbp]
\centering
\includegraphics[width=.6\linewidth]{hiCLIP_data_analysis_1-Select_intra_molecular_hybrids}
\caption{The RNA source of hybrid reads from intra-molecular interactions.}
\label{fig:Select_intra_molecular_hybrids}
\end{figure}

The interacting partners of lncRNAs are analyzed as follows.

<<Search_lncRNA_interacting_partner, echo=TRUE, fig=TRUE, include=FALSE>>=
interGenePartner(inter.gene.grl, "lncRNA")
@

\begin{figure}[htbp]
\centering
\includegraphics[width=.6\linewidth]{hiCLIP_data_analysis_1-Search_lncRNA_interacting_partner}
\caption{Interacting partner of lncRNA.}
\label{fig:Search_lncRNA_interacting_partner}
\end{figure}

\section{Export hybrid read data for data visualization on IGV}

The hybrid reads representing intra-gene RNA structures are exported as sam file, which is compatible with IGV browser. 

<<Export directory, echo=FALSE>>=
high.RNase.hgrL.file <- paste(hiCLIP.dir, "/results/manuscript/IGV/hybrid_highRNase.sam", sep = "")
low.RNase.hgrL.file <- paste(hiCLIP.dir, "/results/manuscript/IGV/hybrid_lowRNase.sam", sep = "")
@

<<Export hybrid data, echo=TRUE>>=
low.gene.grL <- selectGeneMappedHybrid(low.grL)
high.gene.grL <- selectGeneMappedHybrid(high.grL) 

low.intra.gene.grL <- selectHybridType(low.gene.grL, "intra")
high.intra.gene.grL <- selectHybridType(high.gene.grL, "intra")

exportHybrid(high.intra.gene.grL, high.RNase.hgrL.file)
exportHybrid(low.intra.gene.grL, low.RNase.hgrL.file)
@

\section{Folding energy analysis of hybrid reads}

In order to confirm that hybrid reads can identify RNA duplex directly, the folding energy of hybrid reads is calculated using \texttt{RNAhybrid} and compared with randomly re-positioned reads.

<<Set directory for RNAhybrid analysis, echo=FALSE>>=
gr.tc.longest.gene.Rdata <- paste(hiCLIP.dir, "/data/processed/Rdata/gr_tc_longest_gene.Rdata", sep = "")
RNAhybrid.intra.all.dir <-  paste(RNAhybrid.dir, "/intra.gene.all", sep = "")
RNAhybrid.intra.utr3.all.dir <- paste(RNAhybrid.dir, "/intra.utr3.all", sep = "")
RNAhybrid.intra.CDS.all.dir <- paste(RNAhybrid.dir, "/intra.CDS.all", sep = "")
RNAhybrid.inter.all.dir <-  paste(RNAhybrid.dir, "/inter.gene.all", sep = "")
RNAhybrid.intra.inter.all.dir <- paste(RNAhybrid.dir, "/intra.inter.gene.all", sep = "")
RNAhybrid.intra.inter.random.all.dir <- paste(RNAhybrid.dir, "/intra.inter.random.gene.all", sep = "")
longest.gene.fasta <- paste(hiCLIP.dir, "/data/processed/fasta/longest_mRNA_and_ncRNA.fa", sep = "")
@

<<RNAhybridnalysis_intragene_all, echo=TRUE, fig=TRUE, include=FALSE>>=
load(gr.tc.longest.gene.Rdata)
sr.longest.gene <- readFasta(longest.gene.fasta)

RNAhybridAnalysis(
  intra.gene.grl, 
  all.gr = gr.tc.longest.gene, 
  fasta = sr.longest.gene, 
  out.dir = RNAhybrid.intra.all.dir, 
  hiCdir = hiCLIP.dir, 
  pdir = python.dir
  )
@

\begin{figure}[htbp]
\centering
\includegraphics[width=.6\linewidth]{hiCLIP_data_analysis_1-RNAhybridnalysis_intragene_all}
\caption{The folding energy of hybrid reads (Intragene hybrid (all)).}
\label{fig:RNAhybridnalysis_intragene_all}
\end{figure}

<<RNAhybridnalysis_intragene-utr3_all, echo=TRUE, fig=TRUE, include=FALSE>>=
intra.utr3.gene.hgrL <- selectHybridAnnot(
  intra.gene.grl, 
  t.gtf = gr.tc.longest.mRNA,
  annot = "utr3"
  )

RNAhybridAnalysis(
  intra.utr3.gene.hgrL, 
  all.gr = gr.tc.longest.gene, 
  fasta = sr.longest.gene, 
  out.dir = RNAhybrid.intra.utr3.all.dir, 
  hiCdir = hiCLIP.dir, 
  pdir = python.dir
  )
@

\begin{figure}[htbp]
\centering
\includegraphics[width=.6\linewidth]{hiCLIP_data_analysis_1-RNAhybridnalysis_intragene-utr3_all}
\caption{The folding energy of hybrid reads (intragene-utr3 hybrid).}
\label{fig:RNAhybridnalysis_intragene-utr3_all}
\end{figure}

<<RNAhybridnalysis_intragene-CDS_all, echo=TRUE, fig=TRUE, include=FALSE>>=
intra.CDS.gene.hgrL <- selectHybridAnnot(
  intra.gene.grl, 
  t.gtf = gr.tc.longest.mRNA,
  annot = "CDS"
  )

RNAhybridAnalysis(
  intra.CDS.gene.hgrL, 
  all.gr = gr.tc.longest.gene, 
  fasta = sr.longest.gene, 
  out.dir = RNAhybrid.intra.CDS.all.dir, 
  hiCdir = hiCLIP.dir, 
  pdir = python.dir
  )
@

\begin{figure}[htbp]
\centering
\includegraphics[width=.6\linewidth]{hiCLIP_data_analysis_1-RNAhybridnalysis_intragene-CDS_all}
\caption{The folding energy of hybrid reads (intragene-CDS hybrid).}
\label{fig:RNAhybridnalysis_intragene-CDS_all}
\end{figure}

<<RNAhybridnalysis_intergene_all, echo=TRUE, fig=TRUE, include=FALSE>>=
RNAhybridAnalysis(
  inter.gene.grl, 
  all.gr = gr.tc.longest.gene, 
  fasta = sr.longest.gene, 
  out.dir = RNAhybrid.inter.all.dir, 
  hiCdir = hiCLIP.dir, 
  pdir = python.dir
  )
@

\begin{figure}[htbp]
\centering
\includegraphics[width=.6\linewidth]{hiCLIP_data_analysis_1-RNAhybridnalysis_intergene_all}
\caption{The folding energy of hybrid reads (intergene hybrid (all)).}
\label{fig:RNAhybridnalysis_intergene_all}
\end{figure}

<<RNAhybridnalysis_for_all_inter_intra, echo=TRUE, fig=TRUE, include=FALSE>>=

intra.inter.gene.hgrl <- mergeHybrid(intra.gene.grl, inter.gene.grl)

RNAhybridAnalysis(
  intra.inter.gene.hgrl, 
  all.gr = gr.tc.longest.gene, 
  fasta = sr.longest.gene, 
  out.dir = RNAhybrid.intra.inter.all.dir, 
  hiCdir = hiCLIP.dir, 
  pdir = python.dir
  )
@

\begin{figure}[htbp]
\centering
\includegraphics[width=.6\linewidth]{hiCLIP_data_analysis_1-RNAhybridnalysis_for_all_inter_intra}
\caption{The folding energy of all hybrid reads (both intra and inter gene hybrid.}
\label{fig:RNAhybridnalysis_for_all_inter_intra}
\end{figure}

<<Export intra inter gene hgrl data as an Rdata file, echo=FALSE>>=
intra.inter.gene.hgrl.Rdata <- paste(hiCLIP.dir, "/data/processed/Rdata/intra.inter.gene.hgrl.Rdata", sep = "")
save(intra.inter.gene.hgrl, file = intra.inter.gene.hgrl.Rdata) 
@

<<RNAhybridnalysis_for_all_inter_intra_vs_shuffle, echo=TRUE, fig=TRUE, include=FALSE>>=
RNAhybridAnalysis2(
  intra.inter.gene.hgrl, 
  all.gr = gr.tc.longest.gene, 
  fasta = sr.longest.gene, 
  out.dir = RNAhybrid.intra.inter.random.all.dir, 
  hiCdir = hiCLIP.dir, 
  pdir = python.dir
  )
@

\begin{figure}[htbp]
\centering
\includegraphics[width=.6\linewidth]{hiCLIP_data_analysis_1-RNAhybridnalysis_for_all_inter_intra_vs_shuffle}
\caption{The folding energy of all hybrid reads (both intra and inter gene hybrid, but the shuffled sequences were used as the control.}
\label{fig:RNAhybridnalysis_for_all_inter_intra_vs_shuffle}
\end{figure}

\section{Definition of non-redundant set of duplexes}

The RNA duplex identified by at least two independent hybrid reads are defined as confident RNA duplex.

<<Set directory for confident duplex identification, echo=FALSE>>=
distance.dir <- paste(hiCLIP.dir, "/results/manuscript/range", sep = "")
distance.confident.dir <- paste(distance.dir, "/confident", sep = "")

confident.island.file <- paste(hiCLIP.dir, "/results/manuscript/IGV/confident_island.sam", sep = "")
confident.duplex.file <- paste(hiCLIP.dir, "/results/manuscript/IGV/confident_RNA_duplex.sam", sep = "")
@

<<Identifiy confident RNA duplex, echo=TRUE>>=
confident.island.hgrL <- confidentIsland(
  intra.gene.grl, low.val = 2, width.min = 8
  )

confident.island.ext.hgrL <- extendHybrid(
  confident.island.hgrL,
  min.width = 17
  )

confident.duplex.hgrL <- findDuplex2(
  confident.island.ext.hgrL, sr.longest.gene, 
  distance.confident.dir, hiCLIP.dir
  )

analyzeRNADuplexRange(
  confident.duplex.hgrL, 
  t.gtf = gr.tc.longest.mRNA, 
  out.dir =  distance.confident.dir, 
  hiCdir = hiCLIP.dir, 
  print.flag = FALSE
  )

exportHybrid(confident.island.hgrL, confident.island.file)
exportHybrid(confident.duplex.hgrL, confident.duplex.file)

@

<<Save_confident_duplexes, echo=FALSE>>=
confident.duplex.hgrL.Rdata <- paste(hiCLIP.dir, "/data/processed/Rdata/confident.duplex.hgrL.Rdata", sep = "")
save(confident.duplex.hgrL, file = confident.duplex.hgrL.Rdata) 

confident.island.hgrL.Rdata <- paste(hiCLIP.dir, "/data/processed/Rdata/confident.island.hgrL.Rdata", sep = "")
save(confident.island.hgrL, file = confident.island.hgrL.Rdata)
@


<<Set directories for all hybrid read definition, echo=TRUE>>=
intra.gene.hgrL.file <- paste(hiCLIP.dir, "/results/manuscript/IGV/intra_gene_hybrid_all.sam", sep = "")
intra.gene.single.hgrL.file <- paste(hiCLIP.dir, "/results/manuscript/IGV/intra_gene_hybrid_single.sam", sep = "")

distance.single.dir <- paste(distance.dir, "/single", sep = "")

intra.gene.single.duplex.hgrL.file <- paste(hiCLIP.dir, "/results/manuscript/IGV/intra_gene_hybrid_single_duplex.sam", sep = "")

all.collapsed.duplex.hgrL.Rdata <- paste(hiCLIP.dir, "/data/processed/Rdata/all.collapsed.duplex.hgrL.Rdata", sep = "")
all.collapsed.duplex.hgrL.file <- paste(hiCLIP.dir, "/results/manuscript/IGV/all_collapsed_duplex.sam", sep = "")
@

<<Processing data to extract single reads and all reads, echo=TRUE>>=
intra.gene.single.hgrL <- findOverlapHGRL(intra.gene.grl, confident.island.hgrL, type = "subtract")

exportHybrid(intra.gene.grl, intra.gene.hgrL.file)
exportHybrid(intra.gene.single.hgrL, intra.gene.single.hgrL.file)


single.duplex.hgrL <- findDuplex2(
  intra.gene.single.hgrL, sr.longest.gene, 
  distance.single.dir, hiCLIP.dir
  )
  
single.duplex.hgrL <- sortHybrid(single.duplex.hgrL)
single.duplex.hgrL <- collapsedIdenticalHGRL(single.duplex.hgrL)

exportHybrid(single.duplex.hgrL, intra.gene.single.duplex.hgrL.file)

@

All duplexes (= duplexes identified by single reads + duplexes identified by the confident islands) are prepared and later further examined.

<<Prepaing all reads, echo=TRUE>>=
all.collapsed.duplex.hgrL <- mergeHybrid(single.duplex.hgrL, confident.duplex.hgrL)
all.collapsed.duplex.hgrL <- sortHybrid(all.collapsed.duplex.hgrL)

save(all.collapsed.duplex.hgrL, file = all.collapsed.duplex.hgrL.Rdata)

exportHybrid(all.collapsed.duplex.hgrL, all.collapsed.duplex.hgrL.file)

print(
  sprintf("The number of genes containing at least one hybrid read is %d",
  length(unique(seqnames(all.collapsed.duplex.hgrL$L))))
  )
@

\section{RNA source of the all reads}
RNA source of the all identified reads are examined.

<<Load functions for source analysis, echo=FALSE>>=
annotation.tab.file <- paste(hiCLIP.dir, "/data/processed/annotation/gene_annotations.tab", sep = "")
annotation.df <- read.table(annotation.tab.file, stringsAsFactors = FALSE, header = TRUE)

gr.tc.longest.mRNA.Rdata <- paste(hiCLIP.dir, "/data/processed/Rdata/gr_tc_longest_mRNA.Rdata", sep = "")
load(gr.tc.longest.mRNA.Rdata)

gr.mRNA.with.intron.Rdata <- paste(hiCLIP.dir, "/data/processed/Rdata/gr_mRNA_with_intron.Rdata", sep = "")
load(gr.mRNA.with.intron.Rdata)

distance.all.dir <- paste(distance.dir, "/all", sep = "")
@


<<RNA_sorce_all, echo=TRUE, fig=TRUE, include=FALSE>>=
all.collapsed.duplex.hgrL <- 
  addColumnHGRL(all.collapsed.duplex.hgrL, 
                "category", 
                default.value = "NA")

all.collapsed.duplex.hgrL <- 
  addColumnHGRL(all.collapsed.duplex.hgrL, 
                "annot", 
                default.value = "NA")

all.collapsed.duplex.hgrL <- annotateHybrid(
  all.collapsed.duplex.hgrL, 
  annotation.df, 
  gr.tc.longest.mRNA, 
  gr.mRNA.with.intron
  )

plotHybridPairRNASource(all.collapsed.duplex.hgrL)
@

\begin{figure}[htbp]
\centering
\includegraphics[width=.6\linewidth]{hiCLIP_data_analysis_1-RNA_sorce_all}
\caption{The RNA source of the all idnetified duplexes.}
\label{fig:RNA_sorce_all}
\end{figure}
@



\section{Analysis of the loop length of RNA structures.}
 
The loop length of RNA structures in CDS and 3\textprime\ UTR is compared by Mann-Whitney U test.

<<Analyze_RNA_structure_range_all, echo=TRUE, fig=TRUE, include=FALSE, width=5, height=3>>=
analyzeRNADuplexRange(
  all.collapsed.duplex.hgrL, 
  t.gtf = gr.tc.longest.mRNA, 
  out.dir =  distance.all.dir, 
  hiCdir = hiCLIP.dir, 
  print.flag = TRUE
  )
@


\begin{figure}[htbp]
\centering
\includegraphics[width=.6\linewidth]{hiCLIP_data_analysis_1-Analyze_RNA_structure_range_all}
\caption{The range of RNA structures in CDS and 3\textprime\ UTR.}
\label{fig:Analyze_RNA_structure_range_all}
\end{figure}


\section{Session Info}
<<sessinInfo, results=tex, echo=FALSE>>=
toLatex(sessionInfo())

proc.time() - ptm
@

\end{document}
